File: src/main/LevelSpace.java
package org.nlogo.ls;

import scala.collection.JavaConverters;

import java.awt.Component;
import java.awt.Container;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.GraphicsEnvironment;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.net.MalformedURLException;
import java.util.concurrent.ConcurrentHashMap;
import java.util.Map;
import java.util.List;
import java.util.Collections;
import java.util.Set;
import java.util.stream.Collectors;

import javax.swing.*;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

import com.google.common.collect.MapMaker;

import org.nlogo.nvm.HaltException;
import org.nlogo.nvm.ExtensionContext;

import org.nlogo.api.*;
import org.nlogo.app.*;
import org.nlogo.core.ExtensionObject;
import org.nlogo.core.LogoList;
import org.nlogo.core.CompilerException;
import org.nlogo.core.Token;
import org.nlogo.core.Syntax;
import org.nlogo.core.SyntaxJ;
import org.nlogo.awt.EventQueue$;
import org.nlogo.workspace.AbstractWorkspaceScala;
import org.nlogo.window.GUIWorkspace;

import org.nlogo.ls.gui.ModelManager;

public class LevelSpace implements org.nlogo.api.ClassManager {

    // This can be accessed by both the JobThread and EDT (when halting)
    private final Map<Integer, ChildModel> models = new ConcurrentHashMap<Integer, ChildModel>();

    // counter for keeping track of new models
    private int modelCounter = 0;

    public LetPrim letManager = new LetPrim();

    // These need to be cleaned up on unload
    private JMenuItem haltButton;
    private ActionListener haltListener = new ActionListener() {
        @Override
        public void actionPerformed(ActionEvent arg0) {
            haltChildModels(models);
        }
    };

    private LSModelManager modelManager = GraphicsEnvironment.isHeadless() ? new HeadlessBackingModelManager() : new BackingModelManager();

    @Override
    public void load(PrimitiveManager primitiveManager) throws ExtensionException {
        primitiveManager.addPrimitive("let", letManager);
        primitiveManager.addPrimitive("ask", new Ask(this));
        primitiveManager.addPrimitive("of", new Of(this));
        primitiveManager.addPrimitive("report", new Report(this));
        primitiveManager.addPrimitive("with", new With(this));
        primitiveManager.addPrimitive("create-models", new CreateModels<HeadlessChildModel>(HeadlessChildModel.class));
        primitiveManager.addPrimitive("create-interactive-models", new CreateModels<GUIChildModel>(GUIChildModel.class));
        primitiveManager.addPrimitive("name-of", new Name(this));
        primitiveManager.addPrimitive("set-name", new SetName(this));
        primitiveManager.addPrimitive("close", new Close(this));
        primitiveManager.addPrimitive("models", new AllModels(this));
        primitiveManager.addPrimitive("model-exists?", new ModelExists(this));
        primitiveManager.addPrimitive("reset", new Reset());
        primitiveManager.addPrimitive("path-of", new Path(this));
        primitiveManager.addPrimitive("display", new UpdateView(this));
        primitiveManager.addPrimitive("show", new Show(this));
        primitiveManager.addPrimitive("hide", new Hide(this));
        primitiveManager.addPrimitive("show-all", new ShowAll(this));
        primitiveManager.addPrimitive("hide-all", new HideAll(this));
        primitiveManager.addPrimitive("uses-level-space?", new UsesLS(this));


        if (!GraphicsEnvironment.isHeadless()) {
            // Adding event listener to Halt for halting child models
            MenuElement[] elements = App.app().frame().getJMenuBar().getSubElements();
            for (MenuElement e : elements) {
                if (e instanceof ToolsMenu) {
                    ToolsMenu tm = (ToolsMenu) e;
                    haltButton = tm.getItem(0);
                    haltButton.addActionListener(haltListener);
                }
            }
        }
    }

    public boolean isMainModel(ExtensionManager myEM) {
        return myEM == App.app().workspace().getExtensionManager();
    }

    public ChildModel getModel(int id) throws ExtensionException {
        if (models.containsKey(id)) {
            return models.get(id);
        } else {
            throw new ExtensionException("There is no model with ID " + id);
        }
    }

    public boolean containsModel(int id) {
        return models.containsKey(id);
    }

    public Set<Integer> modelSet() {
        return models.keySet();
    }

    public static int castToId(Object id) throws ExtensionException {
        if (id instanceof Number) {
            return ((Number) id).intValue();
        } else {
            throw new ExtensionException("Expected a model ID but got: " + id);
        }
    }


    @Override
    public void unload(ExtensionManager em) throws ExtensionException {
        if (!GraphicsEnvironment.isHeadless() && isMainModel(em)) {
            App.app().frame().getJMenuBar().remove(modelManager.guiComponent());
        }
        if (haltButton != null) {
            haltButton.removeActionListener(haltListener);
        }
        try {
            reset();
        } catch (HaltException e) {
            // we can ignore this
        }
    }


    private static String getModelPath(ExtensionContext ctx, String basePath) throws ExtensionException {
        try {
            return ctx.attachCurrentDirectory(basePath);
        } catch (MalformedURLException e) {
            throw new ExtensionException(e);
        }
    }

    public class CreateModels<T extends ChildModel> implements Command {
        private Class<T> modelType;

        private CreateModels(Class<T> modelType) {
            this.modelType = modelType;
        }

        @Override
        public Syntax getSyntax() {
            return SyntaxJ.commandSyntax(
                    new int[] { Syntax.NumberType(), Syntax.StringType(), Syntax.CommandType() | Syntax.RepeatableType()}, 2);
        }

        @Override
        public void perform(Argument args[], Context ctx) throws ExtensionException, org.nlogo.api.LogoException {
            AbstractWorkspaceScala parentWS = (AbstractWorkspaceScala) ctx.workspace();

            String modelPath = getModelPath((ExtensionContext) ctx, args[1].getString());
            try {
                for (int i=0; i < args[0].getIntValue(); i++) {
                    ChildModel model;
                    if (modelType == HeadlessChildModel.class || GraphicsEnvironment.isHeadless() || parentWS.behaviorSpaceRunNumber() != 0) {
                        model = new HeadlessChildModel((AbstractWorkspaceScala) ctx.workspace(), modelPath, modelCounter);
                    } else {
                        model = new GUIChildModel(LevelSpace.this, (AbstractWorkspaceScala) ctx.workspace(), modelPath, modelCounter);
                        Workspace rootWS = App.app().workspace();
                        if (rootWS instanceof GUIWorkspace) {
                            model.setSpeed(((GUIWorkspace) rootWS).updateManager().speed());
                        }
                    }
                    model.workspace().behaviorSpaceRunNumber(parentWS.behaviorSpaceRunNumber());
                    model.workspace().behaviorSpaceExperimentName(parentWS.behaviorSpaceExperimentName());
                    models.put(modelCounter, model);
                    if (args.length > 2) {
                        args[2].getCommand().perform(ctx, new Object[]{(double) modelCounter});
                    }
                    modelCounter++;
                }
                updateModelMenu();
            } catch (CompilerException e) {
                throw new ExtensionException(modelPath + " has an error in its code: " + e.getMessage(), e);
            } catch (IOException e) {
                throw new ExtensionException("NetLogo couldn't read the file \"" + modelPath + "\". Are you sure it exists and that NetLogo has permission to read it?", e);
            } catch (InterruptedException e) {
                throw new HaltException(false);
            }
        }
    }

    public void updateModelMenu() {
        Runnable reportModelOpened = new Runnable() {
            @Override
            public void run() {
                modelManager.updateChildModels(models);
            }
        };
        EventQueue$.MODULE$.invokeLater(reportModelOpened);
    }

    public void reset() throws ExtensionException, HaltException {
        modelCounter = 0;

        for (ChildModel model : models.values()){
            model.kill();
        }
        models.clear();
    }

    public class Reset implements Command {
        public Syntax getSyntax() {
            return SyntaxJ.commandSyntax();
        }
        public void perform(Argument args[], Context context)
                throws org.nlogo.api.LogoException, ExtensionException {
            reset();
        }
    }

    public ChildModel[] toModelList(Argument arg) throws LogoException, ExtensionException {
        Object obj = arg.get();
        if (obj instanceof Double) {
            return new ChildModel[] { getModel(arg.getIntValue()) };
        } else if (obj instanceof LogoList) {
            LogoList idList = arg.getList();
            ChildModel[] models = new ChildModel[idList.size()];
            int i = 0;
            for (Object modelIdObj : arg.getList().javaIterable()) {
                models[i] = getModel(castToId(modelIdObj));
                i++;
            }
            return models;
        } else {
            throw new ExtensionException("Expected a number or list");
        }
    }

    public void closeModel(ChildModel model) throws ExtensionException, HaltException {
        model.kill();
        models.remove(model.modelID());
        updateModelMenu();
    }

    @Override
    @SuppressWarnings("deprecation")
    public List<String> additionalJars() {
        return null;
    }

    @Override
    public void clearAll() {
        // We want to keep models between clear-alls, yes?
    }

    @Override
    public StringBuilder exportWorld() {
        // Not supported
        return new StringBuilder();
    }

    @Override
    public void importWorld(List<String[]> arg0, ExtensionManager arg1,
                            ImportErrorHandler arg2) throws ExtensionException {
        // Not supported
    }

    @Override
    public ExtensionObject readExtensionObject(ExtensionManager arg0,
                                               String arg1, String arg2) throws ExtensionException,
            CompilerException {
        // Not supported
        return null;
    }

    @Override
    public void runOnce(ExtensionManager em) throws ExtensionException {
        modelManager.updateChildModels(models);

        if (!GraphicsEnvironment.isHeadless() && isMainModel(em)) {
            final JMenuBar menuBar = App.app().frame().getJMenuBar();
            if (menuBar.getComponentIndex(modelManager.guiComponent()) == -1) {
                menuBar.add(modelManager.guiComponent());
            }
        }
    }

    private void haltChildModels(Map<Integer, ChildModel> models){
        for (ChildModel aModel : models.values()){
            aModel.halt();
        }
    }
}

File: src/main/LevelSpace.scala
package org.nlogo.ls

import java.awt.GraphicsEnvironment
import java.awt.event.{ ActionEvent, ActionListener }
import java.lang.{ Double => JDouble }
import java.util
import java.util.Objects
import java.util.concurrent.ConcurrentHashMap
import javax.swing.JMenuItem

import org.nlogo.api.{
  Argument
, DefaultClassManager
, ExtensionException
, ExtensionManager
, ImportErrorHandler
, LogoException
, PrimitiveManager
, Version
}
import org.nlogo.app.{ App, ToolsMenu }
import org.nlogo.awt.EventQueue
import org.nlogo.core.LogoList
import org.nlogo.nvm.HaltException
import org.nlogo.theme.ThemeSync
import org.nlogo.workspace.{ AbstractWorkspace, ExtensionManager => WorkspaceExtensionManager }

import scala.collection.immutable.ArraySeq
import scala.jdk.CollectionConverters.ConcurrentMapHasAsScala

object LevelSpace {
  private var isHeadlessWorkspace = false
  def isHeadless: Boolean =
    isHeadlessWorkspace || GraphicsEnvironment.isHeadless || Objects.equals(System.getProperty("org.nlogo.preferHeadless"), "true")

  @throws[ExtensionException]
  def castToId(id: Any): Int = id match {
    case number: Number => number.intValue
    case _ => throw new ExtensionException("Expected a model ID but got: " + id)
  }

  def checkSupportNetLogoVersion(): Unit = {
    // Due to changes in the tabs "API" of NetLogo desktop, LevelSpace will only work
    // with NetLogo 6.2.0 and newer.  But we want to release updates to the extensions
    // library that will include users on NetLogo 6.1.1 and 6.1.0, for whom this newer
    // LS will *not* work.  So we'll warn them they need to upgrade NetLogo or downgrade
    // LS, rather than just blowing up.  -Jeremy B June 2021
    // This has now been incremented to 7.0.0, due to API changes required for the GUI
    // redesign. (Isaac B 1/25/25)
    val versionErrorMessage =
      """This version of LevelSpace can only be used with NetLogo version 7.0.0 or newer.
      If you updated LevelSpace through the extensions manager, you can use it to uninstall
      the LevelSpace update and go back to the version that came with your NetLogo
      installation.  You can also upgrade NetLogo to the latest version to use this updated
      LevelSpace by going to the CCL website, https://ccl.northwestern.edu/netlogo."""
    def makeVersion(v: String): String = s"NetLogo${if (Version.is3D) " 3D " else " " }$v"
    val minSupportedVersion = Version.numericValue(makeVersion("7.0.0-internal1"))
    val minErrorVersion     = Version.numericValue(makeVersion("6.4.0"))
    val currentVersion      = Version.numericValue(Version.version)
    if (minErrorVersion < currentVersion && currentVersion < minSupportedVersion) {
      throw new ExtensionException(versionErrorMessage)
    }
  }
}

class LevelSpace extends DefaultClassManager with ThemeSync { // This can be accessed by both the JobThread and EDT (when halting)
  LevelSpace.checkSupportNetLogoVersion()

  final private val models = new ConcurrentHashMap[Integer, ChildModel].asScala
  // counter for keeping track of new models
  private var modelCounter = 0
  var letManager = new LetPrim
  // These need to be cleaned up on unload
  private lazy val haltButton: Option[JMenuItem] = if (LevelSpace.isHeadless)
    None
  else
    App.app.frame.getJMenuBar.getSubElements.collectFirst{case tm: ToolsMenu => tm.getItem(0)}

  private val haltListener: ActionListener = (_: ActionEvent) => haltChildModels()
  private var modelManager: LSModelManager = new HeadlessBackingModelManager

  @throws[ExtensionException]
  override def load(primitiveManager: PrimitiveManager): Unit = {
    primitiveManager.addPrimitive("let", letManager)
    primitiveManager.addPrimitive("ask", new Ask(this))
    primitiveManager.addPrimitive("of", new Of(this))
    primitiveManager.addPrimitive("report", new Report(this))
    primitiveManager.addPrimitive("with", new With(this))
    primitiveManager.addPrimitive("create-models", new CreateModels[HeadlessChildModel](this, createHeadlessModel))
    primitiveManager.addPrimitive("create-interactive-models", new CreateModels[GUIChildModel](this, createGUIModel))
    primitiveManager.addPrimitive("name-of", new Name(this))
    primitiveManager.addPrimitive("set-name", new SetName(this))
    primitiveManager.addPrimitive("close", new Close(this))
    primitiveManager.addPrimitive("models", new AllModels(this))
    primitiveManager.addPrimitive("model-exists?", new ModelExists(this))
    primitiveManager.addPrimitive("reset", new Reset(this))
    primitiveManager.addPrimitive("path-of", new Path(this))
    primitiveManager.addPrimitive("show", new Show(this))
    primitiveManager.addPrimitive("hide", new Hide(this))
    primitiveManager.addPrimitive("show-all", new ShowAll(this))
    primitiveManager.addPrimitive("hide-all", new HideAll(this))
    primitiveManager.addPrimitive("uses-level-space?", new UsesLS(this))
    primitiveManager.addPrimitive("random-seed", new RandomSeed(this))
    primitiveManager.addPrimitive("assign", new Assign(this))
    // We need to actually listen for halt actions because gui child models can be running independently on their own
    // job threads if the user is interacting with them.
    haltButton.foreach(_.addActionListener(haltListener))
  }

  def isMainModel(myEM: ExtensionManager): Boolean = myEM eq App.app.workspace.getExtensionManager

  @throws[ExtensionException]
  def getModel(id: Int): ChildModel =
    models.getOrElse(id, throw new ExtensionException("There is no model with ID " + id))

  def containsModel(id: Int): Boolean = models.contains(id)

  def modelList: Seq[Integer] = Seq(ArraySeq.unsafeWrapArray(models.keys.toArray.sorted): _*)

  def numModels: Integer = models.size

  @throws[ExtensionException]
  override def unload(em: ExtensionManager): Unit = {
    if (!LevelSpace.isHeadless && isMainModel(em)) {
      App.app.frame.getJMenuBar.remove(modelManager.guiComponent)
      App.app.removeSyncComponent(this)
    }
    haltButton.foreach(_.removeActionListener(haltListener))
    try reset()
    catch {
      case _: HaltException =>
      // we can ignore this
    }
  }

  private def initModel(parentWS: AbstractWorkspace, model: ChildModel): Unit = {
    model.workspace.behaviorSpaceRunNumber(parentWS.behaviorSpaceRunNumber)
    model.workspace.behaviorSpaceExperimentName(parentWS.behaviorSpaceExperimentName)
    model.workspace.mainRNG.setSeed(parentWS.mainRNG.nextInt())
    models.put(modelCounter, model)
    modelCounter += 1
  }

  private def createModel(
    parentWS: AbstractWorkspace,
    path: String,
    modelCreator: (AbstractWorkspace, String) => ChildModel
  ): ChildModel = try {
    val model = modelCreator(parentWS, path)
    initModel(parentWS, model)
    model
  } catch {
    case e: Exception => throw new ExtensionException(e)
  }

  def createHeadlessModel(parentWS: AbstractWorkspace, path: String): ChildModel =
    createModel(parentWS, path, { (pws, p) =>
      new HeadlessChildModel(pws, p, modelCounter)
    })

  def createGUIModel(parentWS: AbstractWorkspace, path: String): ChildModel =
    if (LevelSpace.isHeadless || parentWS.behaviorSpaceRunNumber != 0) {
      createHeadlessModel(parentWS, path)
    } else {
      createModel(parentWS, path, { (pws, p) =>
        val model = new GUIChildModel(this, parentWS, path, modelCounter)
        model.setSpeed(App.app.workspace.updateManager.speed)
        model
      })
    }

  def updateModelMenu(): Unit = {
    EventQueue.invokeLater { () =>
      modelManager.updateChildModels(models)
    }
  }

  @throws[ExtensionException]
  @throws[HaltException]
  def reset(): Unit = {
    modelCounter = 0
    models.values.foreach(_.kill())
    models.clear()
  }

  @throws[LogoException]
  @throws[ExtensionException]
  def toModelList(arg: Argument): Seq[ChildModel] = {
    arg.get match {
      case x: JDouble => Seq(getModel(x.intValue))
      case l: LogoList => l.map(x => getModel(LevelSpace.castToId(x)))
      case _ => throw new ExtensionException("Expected a number or list")
    }
  }

  @throws[ExtensionException]
  @throws[HaltException]
  def closeModel(model: ChildModel): Unit = {
    model.kill()
    models.remove(model.modelID)
    updateModelMenu()
  }

  @throws[ExtensionException]
  override def importWorld(arg0: util.List[Array[String]], arg1: ExtensionManager, arg2: ImportErrorHandler): Unit = {
    // TODO
  }

  @throws[ExtensionException]
  override def runOnce(em: ExtensionManager): Unit = {
    // "Can't we just check the `org.nlogo.preferHeadless` property?"  Well, kind-of, but
    // it turns out that doesn't get set automatically and there are a lot of ways to run
    // NetLogo models headlessly that "forget" to do it.  It's safer to check if the
    // workspace we're using is headless in addition to checking the property.  -Jeremy B
    // July 2022
    LevelSpace.isHeadlessWorkspace = em.isInstanceOf[WorkspaceExtensionManager] &&
      em.asInstanceOf[WorkspaceExtensionManager].workspace.isInstanceOf[AbstractWorkspace] &&
      em.asInstanceOf[WorkspaceExtensionManager].workspace.asInstanceOf[AbstractWorkspace].isHeadless

    if (!LevelSpace.isHeadless) { modelManager = new BackingModelManager }
    modelManager.updateChildModels(models)
    if (!LevelSpace.isHeadless && isMainModel(em)) {
      val menuBar = App.app.frame.getJMenuBar
      if (menuBar.getComponentIndex(modelManager.guiComponent) == -1) {
        menuBar.add(modelManager.guiComponent)
      }
      App.app.addSyncComponent(this)
    }
  }

  private def haltChildModels(): Unit = models.values.foreach(_.halt())

  def syncTheme(): Unit = {
    modelManager.syncTheme()
  }
}
